web/docs/pipelines.md

```diff
@@ -0,0 +1,216 @@
+---
+sidebar_position: 0
+---
+
+# Pipelines
+
+A Tenzir **pipeline** is a chain of **operators** that represents a dataflow.
+Operators are the atomic building blocks that produce, transform, or consume
+data. Think of them as UNIX or Powershell commands where output from one command
+is input to the next:
+
+![Pipeline Chaining](pipeline-chaining.excalidraw.svg)
+
+There exist three types of operators: **sources** that produce data, **sinks**
+that consume data, and transformations that do both:
+
+![Pipeline Structure](pipeline-structure.excalidraw.svg)
+
+## Polymorphic Operators
+
+Tenzir pipelines make one more distinction: the elements that the operators push
+through the pipeline are *typed*. Every operator has an input and an output
+type:
+
+![Input and Output Types](operator-pieces.excalidraw.svg)
+
+When composing pipelines out of operators, the type of adjacent operators have
+to match. Otherwise the pipeline is malformed. Here's an example pipeline with
+matching operators:
+
+![Typed Pipeline](typed-pipeline.excalidraw.svg)
+
+We call any void-to-void operator sequence a **closed pipeline**. Only closed
+pipelines can execute. If a pipeline does not have a source and sink, it would
+"leak" data. If a pipeline is open, the engine auto-completes a source/sink when
+possible or rejects it.
+
+Zooming out in the above type table makes the operator types apparent:
+
+![Operator Types](operator-types.excalidraw.svg)
+
+In fact, we can define the operator type as a function of its input and output types:
+
+```
+(Input x Output) → {Source, Sink, Transformation}
+```
+
+## Multi-Schema Dataflows
+
+Tenzir dataflows are *multi-schema* in that a single pipeline can work
+with heterogeneous types of events, each of which have a different schemas.
+This allows you, for example, to perform aggregations across multiple events.
+Multi-schema dataflows require automatic schema inference at parse time. Tenzir
+parsers, such as [`json`](/formats/json) support this out of the box.
+
+This behavior is very different from execution engines that only work with
+structured data, where the unit of computation is typically a fixed set of
+tables. Schema-less systems, such as document-oriented databases, offer more
+simplicity, at the cost of performance. In the spectrum of performance and ease
+of use, Tenzir therefore [fills a gap](/why-tenzir):
+
+![Structured vs. Document-Oriented](structured-vs-document-oriented.excalidraw.svg)
+
+:::info Eclectic & Super-structured Data
+[Zed](https://amyousterhout.com/papers/zed_cidr23.pdf) has a type system similar
+to Tenzir, with the difference that Zed associates types *with every single
+value*. Unlike Zed, Tenzir uses a "data frame" abstraction and relies on
+homogeneous Arrow record batches of up to 65,535 rows.
```

Comment:
Mostly an FYI, not sure if we need to rewrite this: The limit of $`2^{16}`$ is arbitrarily chosen by us. Some operators can go above it, actually. With the plans for dynamic rebatching as an under-the-hood optimization we actually should go well above it in some cases.

Reply:
The only limit enforced by Arrow is 2 GiB per array, and that's only because migrating to FlatBuffers64 would be a breaking change. There is basically no limit on record batches themselves, and neither is there a limit on our chunks or table slices. We usually cut at the latest after 64 Ki events, but that's more by convention than a technical necessity.